### ISSUE TITLE
Background disconnect / reconnect strategy

### ISSUE THREAD

**Issue by LucasVanDongen** [davidstump/SwiftPhoenixClient#104]
I have a connection that basically should exist whenever the app is running. iOS will not allow you to run sockets in the background unless you're a VoIP app and basically we're not really doing that. What is the strategy to follow to disconnect and reconnect? We sometimes run into problems with the StarScream layer giving a -1000 protocol layer error when we disconnect when going into background an reconnect again coming to the foreground

**Comment by dsrees** [comment]
@LucasVanDongen In our products we use the AppDelegate methods `applicationDidEnterBackground` and `applicationWillEnterForeground` to disconnect/reconnect our websockets manually. Is this an approach you've already tried? Or are you looking for a way to utilize the reconnect timer at the socket layer?

**Comment by LucasVanDongen** [comment]
We did and it's pretty solid now. However we had to write logic around reconnecting while disconnecting when you open/close the application quickly. Perhaps I can anonimize it and share it for the demo application?

**Comment by dsrees** [comment]
Yeah if you want to open a PR to add logic to the demo, that would be greatly appreciated

**Comment by srikanth-chetu** [comment]
Hi, Can you share the code snippet to handle this?

**Comment by srikanth-chetu** [comment]
> We did and it's pretty solid now. However we had to write logic around reconnecting while disconnecting when you open/close the application quickly. Perhaps I can anonimize it and share it for the demo application?

Hi @LucasVanDongen ,

If possible, can you share the logic of how you handle socket connections in background mode?. It would be helpful for us.

Thanks in advance.

**Comment by dsrees** [comment]
You're going to have a hard time keeping the socket alive while the app is in the background. iOS will kill it at will whenever it wants. If you want to make sure your socket properly reconnects when the app is opened from the background then you'll want to use `UIApplication.willEnterForegroundNotification`. This will have iOS notify your app when it comes back in the foreground and then you can call `Socket.connect()`.


```swift
NotificationCenter.default
      .addObserver(forName: UIApplication.willEnterForegroundNotification,
                   object: nil,
                   queue: .main) { notification in
        // Call your Socket connection logic here
    }
```

I would also recommend that you observe `UIApplication.willResignActiveNotification` to call disconnect your socket when the app enters the background.
